lib/cisco_node_utils/pim.rb

```diff
@@ -47,6 +47,44 @@ def self.feature_enable
       config_set('pim', 'feature')
     end
 
+    def self.pims
+      afis = %w(ipv4) # Add ipv6 later
+      hash_final = {}
+      afis.each do |afi|
+        hash_final[afi] = {}
+        default_vrf = 'default'
+        ranges = config_get('pim', 'all_ssm_ranges', afi: Pim.afi_cli(afi))
+        unless ranges.nil?
+          ranges.each do |range|
+            # Get the RPs under default VRF
+            hash_final[afi][default_vrf] =
+                      Pim.new(afi, default_vrf, false)
+            hash_final[afi][default_vrf].ssm_range = (range)
+          end
+        end
+        # Getting all custom vrfs rp_Addrs"
+        vrf_ids = config_get('vrf', 'all_vrfs')
+        vrf_ids.delete_if { |vrf_id| vrf_id == 'management' }
+        vrf_ids.each do |vrf|
+          get_args = { vrf: @vrf, afi: Pim.afi_cli(afi) }
+          get_args[:vrf] = vrf
+          ranges = config_get('pim', 'all_ssm_ranges', get_args)
+          next if ranges.nil?
+          ranges.each do |addr|
+            hash_final[afi] ||= {}
+            hash_final[afi][vrf] =
+                      Pim.new(afi, vrf, false)
+            hash_final[afi][vrf].ssm_range = (addr)
+          end
+        end
+      end
+      hash_final
+    rescue Cisco::CliError => e
+      # cmd will syntax reject when feature is not enabled
+      raise unless e.clierror =~ /Syntax error/
+      return {}
+    end
+
```

Comment:
Sorry for the delay on this review. I noticed earlier that this change was a problem but I didn't get the chance to come up with a codefix for you until now.

The `self.pims` method above can't be used as is - the problem is that you've tied strictly to 'ssm-range', which means that if I have a different property like 'anycast-rp' then I won't get an object unless I've also got an 'ssm-range' configured.

But that's okay, you were on the right track with this. In this case all you need to do is create an object for each vrf and you're done. I drafted up some test code that works -- it returns what I expect to see and passes your minitest anyway:

```
    # self.pims returns a hash of all current pim objects.
    # There will be one object for each vrf. This method has slightly different
    # behavior than most of the other "get all objects" methods because the
    # pim properties are standalone and do not reside in a "pim" container;
    # therefore, this method simply returns a pim object for each vrf
    # regardless of whether there are any specific pim properties currently
    # defined for the context.
    def self.pims
      afis = %w(ipv4) # TBD: No support for ipv6 at this time
      hash_final = {}
      afis.each do |afi|
        hash_final[afi] = {}
        vrf = 'default'
        hash_final[afi][vrf] = Pim.new(afi, vrf, false)

        # Now the vrf's
        vrf_ids = config_get('vrf', 'all_vrfs')
        vrf_ids.delete_if { |vrf_id| vrf_id == 'management' }
        vrf_ids.each do |vrf|
          hash_final[afi][vrf] = Pim.new(afi, vrf, false)
        end
      end
      hash_final
    rescue Cisco::CliError => e
      # cmd will syntax reject when feature is not enabled
      raise unless e.clierror =~ /Syntax error/
      return {}
    end
```

Also, you can remove the `all_ssm_ranges` entry in the yaml:

```
-all_ssm_ranges:
-  multiple:
-  config_get_token_append : '/^<afi> pim ssm range (.*)$/'
-
```